File: sorting/boj_2110_공유기설치.java
package sorting;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class boj_2110_공유기설치 {	
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int arr[] = new int[N];
		for (int i = 0; i < N ; i++) {
			arr[i] = Integer.parseInt(br.readLine());
		}
		int result = 0;
		Arrays.sort(arr);
		int start = 1;
		int end = arr[N-1] - arr[0];
		while(start <= end) {
			int mid = (start + end) / 2;
			int prevHouse = arr[0];
			int count = 1;
			for (int i = 0; i < N; i++) {
				int distance = arr[i] - prevHouse;
				if(distance >= mid) {
					count++;
					prevHouse = arr[i];
				}
			}
			if(count >= M) {
				start = mid + 1;
				result = mid;
			} else {
				end = mid - 1;
			}
		}
		System.out.println(result);
	}
}

File: sorting/boj_2110_공유기설치.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

long long N, C;
vector<long long> wifi;

int main() {
  cin >> N >> C;
  for (int i = 0; i < N; i++) {
    long long input;
    cin >> input;
    wifi.push_back(input);
  }

  sort(wifi.begin(), wifi.end());

  long long low = 1;
  long long high = wifi[N - 1] - wifi[0];
  long long result = 0;
  while (low <= high) {
    long long mid = (low + high) / 2;
    long long prev = wifi[0];
    int count = 1;
    for (int i = 0; i < N; i++) {
      long long distance = wifi[i] - prev;
      if (distance >= mid) {
        count++;
        prev = wifi[i];
      }
    }
    if (count < C) {
      high = mid - 1;
    } else {
      result = mid;
      low = mid + 1;
    }
  }
  cout << result;
}